From 4ec1d467bd65c8ed957a7489a7461c5388f38489 Mon Sep 17 00:00:00 2001
From: Jeffy Chen <jeffy.chen@rock-chips.com>
Date: Fri, 23 Oct 2020 11:11:11 +0800
Subject: [PATCH 20/40] mpegtsdemux: Create new PCR group for big gap

Currently it would try to reuse the current group and handle the gap by
just adding 500ms to pcroffset.

Signed-off-by: Jeffy Chen <jeffy.chen@rock-chips.com>
---
 gst/mpegtsdemux/mpegtspacketizer.c | 11 +++++++++++
 1 file changed, 11 insertions(+)

diff --git a/gst/mpegtsdemux/mpegtspacketizer.c b/gst/mpegtsdemux/mpegtspacketizer.c
index 91df87f..bf24993 100644
--- a/gst/mpegtsdemux/mpegtspacketizer.c
+++ b/gst/mpegtsdemux/mpegtspacketizer.c
@@ -1982,6 +1982,7 @@ record_pcr (MpegTSPacketizer2 * packetizer, MpegTSPCR * pcrtable,
      *    Initialize current to that group
      */
     GST_DEBUG ("No current window estimator, Checking for group to use");
+create_new_group:
     for (tmp = pcrtable->groups; tmp; tmp = tmp->next) {
       PCROffsetGroup *group = (PCROffsetGroup *) tmp->data;
 
@@ -2054,6 +2055,15 @@ record_pcr (MpegTSPacketizer2 * packetizer, MpegTSPCR * pcrtable,
   if (G_UNLIKELY (corpcr - current->pending[current->last].pcr >
           500 * PCR_MSECOND)) {
     GST_DEBUG ("New PCR more than 500ms away, handling discont");
+
+#if 1
+    /**
+     * Create new discont group instead of reusing the current, otherwise
+     * it will try to handle the gap by just adding 500ms to pcroffset.
+     */
+    _close_current_group (pcrtable);
+    goto create_new_group;
+#else
     /* Take values from current and put them in the current group (closing it) */
     /* Create new group with pcr/offset just after the current group
      * and mark it as a discont */
@@ -2061,6 +2071,7 @@ record_pcr (MpegTSPacketizer2 * packetizer, MpegTSPCR * pcrtable,
     _append_group_values (current->group, current->pending[current->last]);
     _set_current_group (pcrtable, current->group, pcr, offset, TRUE);
     return;
+#endif
   }
 
   if (G_UNLIKELY (corpcr == current->last_value.pcr)) {
-- 
2.20.1

